### PR TITLE
Implement tunneling of the GUI through Jupyter notebook

### LINKED ISSUE #817: IPythonViz does not work in Jupyter Notebook that is not on localhost or accessed via HTTPS
There is probably a number of problems here, but the first is that browser block non-HTTPS content on HTTPS sites (and they should). That means we should start Nengo GUI in SSL mode, but there a few problems with that:

1. How do we detect if we need to start in SSL mode? The Python code is executed in the IPyKernel that has no knowledge of the frontend accessing it (it could be even multiple frontends). Potentially it is possible to provide both SSL and non-SSL services with a single GUI server (especially given PR #864). However usually this requires two different ports and then we still do not know which of the two ports to connect to. On the frontend site `//some-address` should automatically select the right protocol and either port 80 or 443 as used by normal webservers. However, we cannot expect to be able to run Nengo GUI on those ports.
2. Where do we get the SSL certificates from? Ideally we would reuse the ones used for the Jupyter server, but I don't think we can obtain those paths automatically. We could require the user to provide them, but that's inconvenient. It might be possible to auto-generate them, but this requires the user to click through that scary warning about self-signed certificates (and without the user being aware that certificates were generated, they might think of this as suspicious?).

### PR DESCRIPTION
**Pull request by jgosmann**
This PR implement tunneling of connections to the Nengo GUI through the Jupyter notebook server. This allows to use `IPythonViz` in notebooks accessed within an SSL protected Jupyter environment or in notebooks that are not accessed from localhost.

This PR is based on #915 and has the commits from #919 cherry-picked. Furthermore, this PR contains a number of other fixes that could be extracted into separate PRs as those should be merged independent of the tunneling functionality (but are mostly required for the tunneling functionality too):

* 6fbb4c81d1d920f1dce86d0853f196b0b533fc9f: Block JS access to session cookie
* 06b445df17f3409d9f77e52610713e6c72a51556: Use HTTP/1.1 for HTTP responses
* 0bc46f13991f1b73de5151aa481d9bd0f89d2166: Use cookies library supporting newer RFC

Closes #817.

To try it out the Jupyter server extension has to be enabled with:

```bash
jupyter serverextension enable nengo_gui.jupyter
```

Still todo:

- [x] Vendorize the cookies library.
- [x] Error handling in the tunneling request handlers.
- [x] [Favicon should be specified in `link` tag](https://en.wikipedia.org/wiki/Favicon#How_to_use) for full prefix support.
- [x] Kernel shutdown takes a long time due to the Nengo GUI server. Can we fix that?
- [x] Some cleanup of the tunneling code
- [x] Installation instructions.

### CODE REVIEW

**Comment by jgosmann** [comment]
This is now finished. If the server extension is not activated and one tries to use `IPythonViz`, it will show instructions on how to activate the extension.

Given the major changes in the PR, should we consider updating the name of the module itself and/or `IPythonViz`?

File: nengo_gui/static/nengo.js [hunk 1]
```diff
@@ -39,7 +39,9 @@ Nengo.create_websocket = function(uid) {
     } else{
         var ws_proto = 'ws:';
     }
-    var ws_url = ws_proto + '//' + parser.host + '/viz_component?uid=' + uid;
+    var ws_url = ws_proto + '//' + parser.host + window.location.pathname +
+        'viz_component?uid=' + uid;
+    console.log(ws_url)
```
Comment (tcstewar) [inline nengo_gui/static/nengo.js]: I assume this console.log should be removed?  :)

Comment (jgosmann) [reply nengo_gui/static/nengo.js]: yeah

**Comment by jgosmann** [comment]
This should be merged with #923 because of f07406b4b634ee73b65b1c851c642e4451020e7c.

**Comment by jgosmann** [comment]
It would be nice to have this merged.

File: nengo_gui/ipython.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ class ConfigReuseWarning(UserWarning):
     pass
 
 
-class IPythonViz(object):
+class InlineGUI(object):
```
Comment (tcstewar) [inline nengo_gui/ipython.py]: I am very much on-board with renaming this, but we'll definitely want a deprecation system here, for backwards compatibility...

Comment (jgosmann) [reply nengo_gui/ipython.py]: Added deprecation warnings.

**Comment by jgosmann** [comment]
I hope that this is ready for review now. Seems to work in Notebook, Lab, and Binder. Turned out that the Jupyter instance might use a base URL that needs to be added to Nengo GUI's prefix, but we can't determine the base URL for `InlineGUI`. Thus I implemented something that allows setting the base URL from the first incoming request. A little bit of a hack, but seems to be the least hacky way to get things working properly.

**Comment by jgosmann** [comment]
@tcstewar What's the roadmap for merging this PR?

**Comment by tcstewar** [comment]
> @tcstewar What's the roadmap for merging this PR?

It looks like there's no unmerged PRs left this depends on, so this PR seems like something that should be up next.  I guess step 1 would be to rebase it to master....  that said, I'm not going to have a chance to look at it closely this week.....  but that seems like a good way to start.

**Comment by jgosmann** [comment]
Rebased.

**Comment by jgosmann** [comment]
@tcstewar please review

**Pull request merged**